### PR TITLE
Add at-rule-name-newline-after rule

### PR DESCRIPTION
**Pull request by evilebottnawi**
/cc @davidtheclark @jeddy3

### CODE REVIEW

File: src/rules/at-rule-name-newline-after/README.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
+# at-rule-name-newline-after
+
+Require a newline after at-rule names.
+
+```css
+@media (max-width: 600px) {}
```
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/README.md]: Example should show the newline.

File: src/rules/at-rule-name-newline-after/README.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+```
+
+```css
+@charset "UTF-8";
+```
+
+```css
+@charset 
+
+"UTF-8";
```
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/README.md]: So it must be a *single* newline? That should be in the sentences.

File: src/rules/at-rule-name-newline-after/__tests__/index.js [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+  }, {
+    code: "@import\r\n\"x.css\";",
+  }, {
+    code: "@import\n\"x.css\" screen and (orientation:landscape);",
+  }, {
+    code: "@import\r\n\"x.css\" screen and (orientation:landscape);",
+  }, {
+    code: "@import\nurl(\"x.css\");",
+  }, {
+    code: "@importn\r\nurl(\"x.css\");",
```
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: I don't think every single test needs to be repeated with `\r\n`. If you just do a couple to demonstrate that behavior is the same, that will be fine and reduce clutter.

File: src/rules/at-rule-name-newline-after/__tests__/index.js [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+  }, {
+    code: "@media\n(min-width: 700px) and (orientation: landscape)\n{ }",
+  }, {
+    code: "@media\r\n(min-width: 700px) and (orientation: landscape)\n{ }",
+  }, {
+    code: "@media\n(min-width: 700px) and (orientation: landscape)\r\n{ }",
+  }, {
+    code: "@media\r\n(min-width: 700px) and (orientation: landscape)\r\n{ }",
+  }, {
+    code: "@media\n(min-width: 700px)  and (orientation: landscape) { }",
```
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: Why is this test important?

File: src/rules/at-rule-name-newline-after/__tests__/index.js [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+  }, {
+    code: "@unknown\n\"ident\";",
+  }, {
+    code: "@unknown\r\n\"ident\";",
+  }, {
+    code: "@unknown\nident { };",
+  }, {
+    code: "@unknown\r\nident { };",
+  }, {
+    code: "a { color: pink; @crazy-custom-at-rule; }",
```
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: Why is this one ok? What is the exception here? It should be documented, I think.

File: src/rules/at-rule-name-newline-after/README.md (hunk 2 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/README.md]: @davidtheclark don't undestand, remove this example?

File: src/rules/at-rule-name-newline-after/__tests__/index.js [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+  } ],
```
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: What if there is an end of line comment:

```css
@media /* my special at-rule */
  print { }
```

File: src/rules/at-rule-name-newline-after/__tests__/index.js (hunk 4 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: @davidtheclark i can remove, but may be false positive if somebody rewrite rule on other logic, be good tested this.

File: src/rules/at-rule-name-newline-after/__tests__/index.js [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+    message: messages.expectedAfter("@media"),
+    line: 1,
+    column: 6,
+  }, {
+    code: "@media(min-width: 700px)\nand (orientation: landscape) { }",
+    message: messages.expectedAfter("@media"),
+    line: 1,
+    column: 6,
+  }, {
+    code: "@media(min-width: 700px)\r\nand (orientation: landscape) { }",
```
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: I'm not sure this set of tests, with newlines before `and`, adds anything to the test suite. Why might this ever fail?

File: src/rules/at-rule-name-newline-after/__tests__/index.js [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+  skipBasicChecks: true,
+
+  accept: [ {
+    code: "@mixin\nmixin() { @content; }; .colors { @include\nmixin { color: $color; }}",
+    description: "ignore content blocks",
+  }, {
+    code: "@mixin\r\nmixin() { @content; }; .colors { @include\r\nmixin { color: $color; }}",
+    description: "ignore content blocks",
+  } ],
+})
```
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: Nice job anticipating the Less and SCSS exceptions.

File: src/rules/at-rule-name-newline-after/__tests__/index.js (hunk 7 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: @davidtheclark agree. remove this lines.

File: src/rules/at-rule-name-newline-after/__tests__/index.js [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+  }, {
+    code: "@media (min-width: 700px) and (orientation: landscape) { }",
+  }, {
+    code: "@media (min-width: 700px) and (orientation: landscape)  { }",
+  }, {
+    code: "@media (min-width: 700px) and (orientation: landscape)\n{ }",
+  }, {
+    code: "@media (min-width: 700px) and (orientation: landscape)\r\n{ }",
+  }, {
+    code: "@media (min-width: 700px) and (orientation: landscape) { }",
```
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: Isn't this a duplicate test?

File: src/rules/at-rule-name-newline-after/__tests__/index.js (hunk 6 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: @davidtheclark good remark

File: src/rules/at-rule-name-newline-after/__tests__/index.js (hunk 5 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: @davidtheclark also need docs in `at-rule-semicolon-space-after`

File: src/rules/at-rule-name-newline-after/__tests__/index.js (hunk 3 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: @davidtheclark agree.

File: src/rules/at-rule-name-newline-after/README.md (hunk 2 above)
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/README.md]: What's the behavior of other `-newline-` rules? Are they always expecting a single line?

File: src/rules/at-rule-name-newline-after/__tests__/index.js (hunk 9 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: @davidtheclark yep remove

File: src/rules/at-rule-name-newline-after/README.md (hunk 2 above)
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/README.md]: I mean, from this and the tests it looks like the rule requires a *single* newline, disallows *multiple* newlines. Two questions, then:

- Is that consistent with other rules requiring newlines?
- Shouldn't that be documented?

File: src/rules/at-rule-name-newline-after/__tests__/index.js (hunk 4 above)
Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: The question, though, is whether the rule's logic would ever look at this area. How is this different from writing tests to make sure the logic doesn't complain about declarations, or selectors? There are an unlimited number of possible ways anything *could* go wrong, so we need to restrict our tests to those that *might* go wrong, or we will lose our minds :)

**Comment by davidtheclark** [comment]
Awesome bunch of PRs this morning, @evilebottnawi.

File: src/rules/at-rule-name-newline-after/README.md (hunk 2 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/README.md]: @davidtheclark I was a bit surprised by this behavior and whether it is normal. That do you think?

File: src/rules/at-rule-name-newline-after/__tests__/index.js (hunk 4 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/__tests__/index.js]: @davidtheclark agree, remove.

File: src/rules/at-rule-name-newline-after/README.md (hunk 2 above)
Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/README.md]: @davidtheclark ping

Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/README.md]: I guess do whatever is consistent with the existing rules, and create a new issue if we need to reconsider.

Comment (evilebottnawi) [inline src/rules/at-rule-name-newline-after/README.md]: @davidtheclark mean all ok here?

Comment (davidtheclark) [inline src/rules/at-rule-name-newline-after/README.md]: Sure.

**Comment by jeddy3** [comment]
@evilebottnawi Just looks like a bit of clean-up is needed before this can be merged.

**Comment by evilebottnawi** [comment]
@jeddy3 I will try to complete all of Monday

**Comment by jeddy3** [comment]
> @jeddy3 I will try to complete all of Monday

There is no rush as we won't be able to review for a couple of weeks.

**Comment by evilebottnawi** [comment]
@jeddy3 ok

**Comment by davidtheclark** [comment]
Merged in beac986b1f250b060dcf16075d7f1083e829215a

**Pull request closed without merging**